streams/actions/src/main/java/com/hortonworks/streamline/streams/actions/topology/state/TopologyStates.java

```diff
@@ -0,0 +1,199 @@
+/**
+ * Copyright 2017 Hortonworks.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+
+ *   http://www.apache.org/licenses/LICENSE-2.0
+
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+package com.hortonworks.streamline.streams.actions.topology.state;
+
+import com.hortonworks.streamline.streams.actions.TopologyActions;
+import com.hortonworks.streamline.streams.catalog.CatalogToLayoutConverter;
+import com.hortonworks.streamline.streams.catalog.Topology;
+import com.hortonworks.streamline.streams.layout.component.TopologyDag;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The class captures the different states a 'topology' can be in and the state transitions.
+ */
+public final class TopologyStates {
```

Comment:
OK. It seems other invalid operations of current state throw IllegalStateException.

Cause of confusion for me is different states operations are pushed into `TopologyState` class. I would have preferred to have `TopologyState` class not containing states operations but method like `execute/run(TopologyContext)` and the implementation would take care of whether the earlier state is valid or not to execute the current state. With that, no need to add any method in TopologyState whenever a new state is added in the existing state flow but other dependent states need to modify accordingly for the validation.

I am fine with the current approach also as we are not going to add states often and change the implementation.